app/controllers/sessions_controller.rb

```diff
@@ -3,14 +3,24 @@ def new
   end
 
   def create
-    user = User.find_by_email(params[:session][:email])
-    if user && user.authenticate(params[:session][:password])
-      session[:user_id] = user.id
-      flash[:success] = "Welcome, #{user.full_name}."
+    if params[:session]
+      user = User.find_by_email(params[:session][:email])
+      if user && user.authenticate(params[:session][:password])
+        session[:user_id] = user.id
+        flash[:success] = "Welcome, #{user.full_name}."
+      else
+        flash[:danger] = "Invalid login."
+      end
     else
-      flash[:danger] = 'Invalid login.'
+      user = User.find_or_create_from_auth(request.env["omniauth.auth"])
+      if user
+        session[:user_id] = user.id
+        flash[:success] = "Welcome, #{user.full_name}."
+      else
+        flash[:danger] = "Invalid login."
+      end
     end
-    redirect_to(:back)
+    redirect_to items_path
```

Comment:
@indiesquidge  because when users login with twitter it takes them to a twitter page, so it doesn't have to be items, but we can't send them back to the twitter page as it's a different site.